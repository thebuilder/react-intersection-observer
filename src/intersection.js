// @flow
import invariant from 'invariant'

type Callback = (inView: boolean, intersectionRatio: number) => void

type Instance = {
  callback: Callback,
  visible: boolean,
  options: IntersectionObserverOptions,
  observerId: ?string,
  observer: ?IntersectionObserver,
}

const INSTANCE_MAP: Map<HTMLElement, Instance> = new Map()
const OBSERVER_MAP: Map<?string, IntersectionObserver> = new Map()

/**
 * Monitor element, and trigger callback when element becomes visible
 * @param element {HTMLElement}
 * @param callback {Function} Called with inView
 * @param options {Object} InterSection observer options
 * @param options.threshold {Number} Number between 0 and 1, indicating how much of the element should be visible before triggering
 * @param options.root {HTMLElement} It should have a unique id or data-intersection-id in order for the Observer to reused.
 * @param options.rootMargin {String} The CSS margin to apply to the root element.
 * @param rootId {String} Unique identifier for the root element, to enable reusing the IntersectionObserver
 */
export function observe(
  element: HTMLElement,
  callback: Callback,
  options: IntersectionObserverOptions = {},
  rootId?: string,
) {
  // Validate that the element is not being used in another <Observer />
  invariant(
    !INSTANCE_MAP.has(element),
    "react-intersection-observer: Trying to observe %s, but it's already being observed by another instance.\nMake sure the `ref` is only used by a single <Observer /> instance.\n\n%s",
    element,
  )
  // IntersectionObserver needs a threshold to trigger, so set it to 0 if it's not defined.
  // Modify the options object, since it's used in the onChange handler.
  if (!options.threshold) options.threshold = 0

  const { root, rootMargin, threshold } = options
  if (!element || !callback) return
  let observerId = rootMargin
    ? `${threshold.toString()}_${rootMargin}`
    : threshold.toString()

  if (root) {
    observerId = rootId ? `${rootId}_${observerId}` : null
  }

  let observerInstance = observerId ? OBSERVER_MAP.get(observerId) : null
  if (!observerInstance) {
    observerInstance = new IntersectionObserver(onChange, options)
    if (observerId) OBSERVER_MAP.set(observerId, observerInstance)
  }

  const instance: Instance = {
    callback,
    visible: false,
    options,
    observerId,
    observer: !observerId ? observerInstance : undefined,
  }

  INSTANCE_MAP.set(element, instance)

  observerInstance.observe(element)

  return instance
}

/**
 * Stop observing an element. If an element is removed from the DOM or otherwise destroyed,
 * make sure to call this method.
 * @param element {HTMLElement}
 */
export function unobserve(element: ?HTMLElement) {
  if (!element) return
  const instance = INSTANCE_MAP.get(element)

  if (instance) {
    const { observerId, observer } = instance
    const observerInstance = observerId
      ? OBSERVER_MAP.get(observerId)
      : observer

    if (observerInstance) {
      observerInstance.unobserve(element)
    }

    // Check if we are still observing any elements with the same threshold.
    let itemsLeft = false
    if (observerId) {
      INSTANCE_MAP.forEach((item, key) => {
        if (item && item.observerId === observerId && key !== element) {
          itemsLeft = true
        }
      })
    }

    if (observerInstance && !itemsLeft) {
      // No more elements to observe for threshold, disconnect observer
      observerInstance.disconnect()
      OBSERVER_MAP.delete(observerId)
    }

    // Remove reference to element
    INSTANCE_MAP.delete(element)
  }
}

/**
 * Destroy all IntersectionObservers currently connected
 **/
export function destroy() {
  OBSERVER_MAP.forEach(observer => {
    observer.disconnect()
  })

  OBSERVER_MAP.clear()
  INSTANCE_MAP.clear()
}

function onChange(changes) {
  changes.forEach(intersection => {
    const { isIntersecting, intersectionRatio, target } = intersection
    const instance = INSTANCE_MAP.get(target)

    // Firefox can report a negative intersectionRatio when scrolling.
    if (instance && intersectionRatio >= 0) {
      const options = instance.options

      let inView = false

      if (Array.isArray(options.threshold)) {
        // If threshold is an array, check if any of them intersects. This just triggers the onChange event multiple times.
        inView = options.threshold.some(threshold => {
          return instance.visible
            ? intersectionRatio > threshold
            : intersectionRatio >= threshold
        })
      } else if (options.threshold !== undefined) {
        // Trigger on 0 ratio only when not visible. This is fallback for browsers without isIntersecting support
        inView = instance.visible
          ? intersectionRatio > options.threshold
          : intersectionRatio >= options.threshold
      }

      if (isIntersecting !== undefined) {
        // If isIntersecting is defined, ensure that the element is actually intersecting.
        // Otherwise it reports a threshold of 0
        inView = inView && isIntersecting
      }

      instance.visible = inView
      instance.callback(inView, intersectionRatio)
    }
  })
}

export default {
  observe,
  unobserve,
  destroy,
}
