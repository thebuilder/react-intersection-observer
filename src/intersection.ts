import invariant from 'invariant'
import { ObserverInstance, ObserverInstanceCallback } from './index'

const INSTANCE_MAP: Map<Element, ObserverInstance> = new Map()
const OBSERVER_MAP: Map<string, IntersectionObserver> = new Map()
const ROOT_IDS: Map<Element, string> = new Map()

let consecutiveRootId = 0

/**
 * Generate a unique ID for the root element
 * @param root
 */
function getRootId(root?: Element | null) {
  if (!root) return ''
  if (ROOT_IDS.has(root)) return ROOT_IDS.get(root)
  consecutiveRootId += 1
  ROOT_IDS.set(root, consecutiveRootId.toString())
  return ROOT_IDS.get(root) + '_'
}

/**
 * Monitor element, and trigger callback when element becomes inView
 * @param element {HTMLElement}
 * @param callback {Function} Called with inView
 * @param options {Object} InterSection observer options
 * @param options.threshold {Number} Number between 0 and 1, indicating how much of the element should be inView before triggering
 * @param options.root {HTMLElement}
 * @param options.rootMargin {String} The CSS margin to apply to the root element.
 */
export function observe(
  element: Element,
  callback: ObserverInstanceCallback,
  options: IntersectionObserverInit = {},
) {
  // IntersectionObserver needs a threshold to trigger, so set it to 0 if it's not defined.
  // Modify the options object, since it's used in the onChange handler.
  if (!options.threshold) options.threshold = 0
  const { root, rootMargin, threshold } = options
  // Validate that the element is not being used in another <Observer />
  invariant(
    !INSTANCE_MAP.has(element),
    "react-intersection-observer: Trying to observe %s, but it's already being observed by another instance.\nMake sure the `ref` is only used by a single <Observer /> instance.\n\n%s",
    element,
  )
  /* istanbul ignore if */
  if (!element) return
  // Create a unique ID for this observer instance, based on the root, root margin and threshold.
  // An observer with the same options can be reused, so lets use this fact
  let observerId: string =
    getRootId(root) +
    (rootMargin
      ? `${threshold.toString()}_${rootMargin}`
      : threshold.toString())

  let observerInstance = OBSERVER_MAP.get(observerId)
  if (!observerInstance) {
    observerInstance = new IntersectionObserver(onChange, options)
    /* istanbul ignore else  */
    if (observerId) OBSERVER_MAP.set(observerId, observerInstance)
  }

  const instance: ObserverInstance = {
    callback,
    element,
    inView: false,
    observerId,
    observer: observerInstance,
    // Make sure we have the thresholds value. It's undefined on a browser like Chrome 51.
    thresholds:
      observerInstance.thresholds ||
      (Array.isArray(threshold) ? threshold : [threshold]),
  }

  INSTANCE_MAP.set(element, instance)
  observerInstance.observe(element)

  return instance
}

/**
 * Stop observing an element. If an element is removed from the DOM or otherwise destroyed,
 * make sure to call this method.
 * @param element {Element}
 */
export function unobserve(element: Element | null) {
  if (!element) return
  const instance = INSTANCE_MAP.get(element)

  if (instance) {
    const { observerId, observer } = instance
    const { root } = observer

    observer.unobserve(element)

    // Check if we are still observing any elements with the same threshold.
    let itemsLeft = false
    // Check if we still have observers configured with the same root.
    let rootObserved = false
    /* istanbul ignore else  */
    if (observerId) {
      INSTANCE_MAP.forEach((item, key) => {
        if (key !== element) {
          if (item.observerId === observerId) {
            itemsLeft = true
            rootObserved = true
          }
          if (item.observer.root === root) {
            rootObserved = true
          }
        }
      })
    }
    if (!rootObserved && root) ROOT_IDS.delete(root)
    if (observer && !itemsLeft) {
      // No more elements to observe for threshold, disconnect observer
      observer.disconnect()
    }

    // Remove reference to element
    INSTANCE_MAP.delete(element)
  }
}

/**
 * Destroy all IntersectionObservers currently connected
 **/
export function destroy() {
  OBSERVER_MAP.forEach(observer => {
    observer.disconnect()
  })

  OBSERVER_MAP.clear()
  INSTANCE_MAP.clear()
  ROOT_IDS.clear()
  consecutiveRootId = 0
}

function onChange(changes: IntersectionObserverEntry[]) {
  changes.forEach(intersection => {
    const { isIntersecting, intersectionRatio, target } = intersection
    const instance = INSTANCE_MAP.get(target)

    // Firefox can report a negative intersectionRatio when scrolling.
    /* istanbul ignore else */
    if (instance && intersectionRatio >= 0) {
      // If threshold is an array, check if any of them intersects. This just triggers the onChange event multiple times.
      let inView = instance.thresholds.some(threshold => {
        return instance.inView
          ? intersectionRatio > threshold
          : intersectionRatio >= threshold
      })

      if (isIntersecting !== undefined) {
        // If isIntersecting is defined, ensure that the element is actually intersecting.
        // Otherwise it reports a threshold of 0
        inView = inView && isIntersecting
      }

      instance.inView = inView
      instance.callback(inView, intersection)
    }
  })
}

export default {
  observe,
  unobserve,
  destroy,
}
