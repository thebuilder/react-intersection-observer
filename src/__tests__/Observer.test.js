/* eslint-disable import/no-named-as-default-member */
import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { observe, unobserve } from '../intersection'
import Observer from '../'
import invariant from 'invariant'

jest.mock('../intersection')
jest.mock('invariant')

const plainChild = ({ ref }) => <div ref={ref} />
afterEach(() => {
  observe.mockReset()
  unobserve.mockReset()
  invariant.mockReset()
})

it('Should render <Observer />', () => {
  const callback = jest.fn(plainChild)
  render(<Observer>{callback}</Observer>)
  expect(callback).toHaveBeenLastCalledWith(
    expect.objectContaining({ inView: false }),
  )
})

it('should render plain children', () => {
  const { container, getByText } = render(<Observer>inner</Observer>)
  getByText('inner')
  const tagName = container.firstChild.tagName.toLowerCase()
  expect(tagName).toBe('div')
})

it('should render with tag', () => {
  const { container, getByText } = render(<Observer tag="span">inner</Observer>)
  getByText('inner')
  const tagName = container.firstChild.tagName.toLowerCase()
  expect(tagName).toBe('span')
})

it('should render with className', () => {
  const { container } = render(
    <Observer className="inner-class">inner</Observer>,
  )
  expect(container.firstChild).toHaveClass('inner-class')
})

it('Should render <Observer /> inview', () => {
  const callback = jest.fn(plainChild)
  observe.mockImplementation((el, callback, options) => {
    callback(true, {})
  })
  render(<Observer>{callback}</Observer>)
  expect(callback).toHaveBeenLastCalledWith(
    expect.objectContaining({ inView: true }),
  )
})

it('Should not render <Observer /> render outside view', () => {
  const { getByText } = render(
    <Observer>
      {({ inView, ref }) => <div ref={ref}>Inview: {inView.toString()}</div>}
    </Observer>,
  )
  getByText('Inview: false')
})

it('Should render <Observer /> render when in view', () => {
  observe.mockImplementation((el, callback, options) => {
    callback(true, {})
  })
  const { getByText } = render(
    <Observer>
      {({ inView, ref }) => <div ref={ref}>Inview: {inView.toString()}</div>}
    </Observer>,
  )

  getByText('Inview: true')
})

it('Should unobserve old node', () => {
  const { rerender, container } = render(
    <Observer>
      {({ inView, ref }) => (
        <div key="1" ref={ref}>
          Inview: {inView.toString()}
        </div>
      )}
    </Observer>,
  )
  unobserve.mockReset()
  observe.mockReset()
  rerender(
    <Observer>
      {({ inView, ref }) => (
        <div key="2" ref={ref}>
          Inview: {inView.toString()}
        </div>
      )}
    </Observer>,
  )
  expect(unobserve).toHaveBeenCalledWith(container.firstChild)
  expect(observe).toHaveBeenCalled()
})

it('Should ensure node exists before observing and unobserving', () => {
  unobserve.mockReset()
  const { unmount } = render(<Observer>{() => null}</Observer>)
  expect(observe).not.toHaveBeenCalled()
  unmount()
  expect(unobserve).not.toHaveBeenCalled()
})

it('Should recreate observer when threshold change', () => {
  const { container, rerender } = render(<Observer>{plainChild}</Observer>)
  observe.mockReset()
  rerender(<Observer threshold={0.5}>{plainChild}</Observer>)
  expect(unobserve).toHaveBeenCalledWith(container.firstChild)
  expect(observe).toHaveBeenCalled()
})

it('Should recreate observer when root change', () => {
  const { container, rerender } = render(<Observer>{plainChild}</Observer>)
  observe.mockReset()
  rerender(<Observer root={{}}>{plainChild}</Observer>)
  expect(unobserve).toHaveBeenCalledWith(container.firstChild)
  expect(observe).toHaveBeenCalled()
})

it('Should recreate observer when rootMargin change', () => {
  const { container, rerender } = render(<Observer>{plainChild}</Observer>)
  observe.mockReset()
  rerender(<Observer rootMargin="10px">{plainChild}</Observer>)
  expect(unobserve).toHaveBeenCalledWith(container.firstChild)
  expect(observe).toHaveBeenCalled()
})

it('Should trigger onChange callback', () => {
  let onObserve
  observe.mockImplementation((el, callback, options) => (onObserve = callback))
  const onChange = jest.fn()
  render(<Observer onChange={onChange}>{plainChild}</Observer>)
  if (onObserve) onObserve(true, {})
  expect(onChange).toHaveBeenLastCalledWith(true, {})
  if (onObserve) onObserve(false, {})
  expect(onChange).toHaveBeenLastCalledWith(false, {})
})

it('Should unobserve when triggerOnce comes into view', () => {
  observe.mockImplementation((el, callback) => {
    if (callback) callback(true, {})
  })
  render(<Observer triggerOnce>{plainChild}</Observer>)
  expect(unobserve).toHaveBeenCalled()
})

it('Should unobserve when unmounted', () => {
  const { container, unmount } = render(<Observer>{plainChild}</Observer>)
  unmount()
  expect(unobserve).toHaveBeenCalledWith(container)
})

it('Should throw error when not passing ref', () => {
  render(
    <Observer>
      {({ inView }) => <div>Inview: {inView.toString()}</div>}
    </Observer>,
  )
  expect(invariant).toHaveBeenLastCalledWith(
    null,
    'react-intersection-observer: No DOM node found. Make sure you forward "ref" to the root DOM element you want to observe.',
  )
})

it('plain children should not catch bubbling onChange event', () => {
  const onChange = jest.fn()
  const { getByRole } = render(
    <Observer onChange={onChange}>
      <input name="field" />
    </Observer>,
  )
  const input = getByRole('textbox')
  fireEvent.change(input, { target: { value: 'changed value' } })
  expect(onChange).not.toHaveBeenCalled()
})
