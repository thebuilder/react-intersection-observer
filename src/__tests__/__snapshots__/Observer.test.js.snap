// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not render <Observer /> render outside view 1`] = `
<InView
  threshold={0}
  triggerOnce={false}
>
  <div>
    Inview: 
    false
  </div>
</InView>
`;

exports[`Should render <Observer /> render when in view 1`] = `
<InView
  threshold={0}
  triggerOnce={false}
>
  <div>
    Inview: 
    true
  </div>
</InView>
`;

exports[`should render plain children 1`] = `
<InView
  threshold={0}
  triggerOnce={false}
>
  <div>
    inner
  </div>
</InView>
`;

exports[`should render with className 1`] = `
<InView
  className="inner-class"
  threshold={0}
  triggerOnce={false}
>
  <div
    className="inner-class"
  >
    inner
  </div>
</InView>
`;

exports[`should render with tag 1`] = `
<InView
  tag="span"
  threshold={0}
  triggerOnce={false}
>
  <span>
    inner
  </span>
</InView>
`;
